File: frontend/src/api/camera.ts
/**
 * 摄像头管理 API
 */
import { request, type ApiResponse, type PageResponse } from './request'

// 摄像头状态
export type CameraStatus = 'online' | 'offline' | 'error'

// 摄像头信息
export interface Camera {
  id: string
  name: string
  code: string | null
  description: string | null
  area_id: string | null
  area_name?: string
  rtsp_url: string
  manufacturer: string | null
  device_model: string | null
  ip_address: string | null
  location: string | null
  longitude: number | null
  latitude: number | null
  is_enabled: boolean
  status: CameraStatus
  /** 在线状态（来自 Redis camera:online:{id}） */
  online?: boolean
  fps: number
  resolution: string | null
  /** 识别间隔(秒)，控制该摄像头推理抽帧频率 */
  inference_interval_sec: number
  algorithm_count?: number
  inference_started?: boolean
  snapshot_url?: string | null
  created_at: string
  updated_at: string
}

// 创建摄像头请求
export interface CreateCameraRequest {
  name: string
  code?: string
  description?: string
  area_id?: string
  rtsp_url: string
  manufacturer?: string
  device_model?: string
  ip_address?: string
  location?: string
  longitude?: number
  latitude?: number
  is_enabled?: boolean
  fps?: number
  resolution?: string
  inference_interval_sec?: number
}

// 更新摄像头请求
export interface UpdateCameraRequest {
  name?: string
  code?: string
  description?: string
  area_id?: string
  rtsp_url?: string
  manufacturer?: string
  device_model?: string
  ip_address?: string
  location?: string
  longitude?: number
  latitude?: number
  is_enabled?: boolean
  fps?: number
  resolution?: string
  inference_interval_sec?: number
}

// 摄像头列表查询参数
export interface CameraQueryParams {
  page?: number
  page_size?: number
  area_id?: string
  status?: CameraStatus
  is_enabled?: boolean
  keyword?: string
}

// 播放 URL
export interface PlayUrls {
  rtsp: string | null
  rtmp: string | null
  http_flv: string | null
  hls: string | null
  ws_flv: string | null
}

/**
 * 获取摄像头列表
 */
export function getCameraList(params?: CameraQueryParams) {
  return request.get<PageResponse<Camera>>('/cameras', { params })
}

/**
 * 获取摄像头详情
 */
export function getCamera(id: string) {
  return request.get<Camera>(`/cameras/${id}`)
}

/**
 * 创建摄像头
 */
export function createCamera(data: CreateCameraRequest) {
  return request.post<Camera>('/cameras', data)
}

/**
 * 更新摄像头
 */
export function updateCamera(id: string, data: UpdateCameraRequest) {
  return request.put<Camera>(`/cameras/${id}`, data)
}

/**
 * 删除摄像头
 */
export function deleteCamera(id: string) {
  return request.delete(`/cameras/${id}`)
}

/**
 * 启动摄像头
 */
export function startCamera(id: string) {
  return request.post(`/cameras/${id}/start`)
}

/**
 * 停止摄像头
 */
export function stopCamera(id: string) {
  return request.post(`/cameras/${id}/stop`)
}

/**
 * 启动摄像头推理（后台推理）
 */
export function startCameraInference(id: string) {
  return request.post(`/cameras/${id}/start-inference`)
}

/**
 * 停止摄像头推理（后台推理）
 */
export function stopCameraInference(id: string) {
  return request.post(`/cameras/${id}/stop-inference`)
}

/**
 * 摄像头直播心跳（播放时每 60 秒调用一次）
 */
export function cameraLiveHeartbeat(id: string) {
  return request.post(`/cameras/${id}/live-heartbeat`)
}

/**
 * 获取播放地址
 */
export function getCameraPlayUrls(id: string) {
  return request.get<PlayUrls>(`/cameras/${id}/play-url`)
}

/** 流通性测试请求 */
export interface ProbeStreamRequest {
  rtsp_url: string
}

/** 流通性测试响应 */
export interface ProbeStreamResult {
  width: number
  height: number
  fps: number | null
  resolution: string | null
}

/**
 * 流通性测试（后端用 OpenCV/ffprobe 获取宽高、帧率）
 */
export function probeStream(data: ProbeStreamRequest) {
  return request.post<ProbeStreamResult>('/cameras/probe-stream', data)
}

/**
 * 抓拍并保存为最新缩略图
 */
export function snapshotCamera(id: string) {
  return request.post<{ snapshot_url: string }>(`/cameras/${id}/snapshot`)
}

File: backend/app/schemas/camera.py
# -*- coding: utf-8 -*-
"""
摄像头 Schema

定义摄像头相关的请求和响应模式
"""
from typing import Optional, List
from datetime import datetime

from pydantic import BaseModel, Field


class CameraBase(BaseModel):
    """
    摄像头基础字段
    """
    
    name: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="摄像头名称"
    )
    code: Optional[str] = Field(
        default=None,
        max_length=50,
        description="摄像头编码"
    )
    description: Optional[str] = Field(
        default=None,
        description="摄像头描述"
    )
    area_id: Optional[str] = Field(
        default=None,
        description="所属区域ID"
    )
    rtsp_url: str = Field(
        ...,
        max_length=500,
        description="RTSP 流地址（可含认证，如 rtsp://user:pass@host/path）"
    )
    manufacturer: Optional[str] = Field(
        default=None,
        max_length=50,
        description="设备厂商"
    )
    device_model: Optional[str] = Field(
        default=None,
        max_length=50,
        description="设备型号"
    )
    ip_address: Optional[str] = Field(
        default=None,
        max_length=50,
        description="设备IP"
    )
    location: Optional[str] = Field(
        default=None,
        max_length=200,
        description="安装位置"
    )
    longitude: Optional[float] = Field(
        default=None,
        description="经度"
    )
    latitude: Optional[float] = Field(
        default=None,
        description="纬度"
    )
    fps: int = Field(
        default=25,
        ge=1,
        le=60,
        description="帧率"
    )
    inference_interval_sec: int = Field(
        default=5,
        ge=1,
        le=3600,
        description="识别间隔(秒)，控制该摄像头推理抽帧频率"
    )
    resolution: Optional[str] = Field(
        default=None,
        max_length=20,
        description="分辨率"
    )


class CameraCreate(CameraBase):
    """
    创建摄像头请求
    """
    
    is_enabled: bool = Field(
        default=True,
        description="是否启用"
    )


class CameraUpdate(BaseModel):
    """
    更新摄像头请求
    """
    
    name: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=100,
        description="摄像头名称"
    )
    code: Optional[str] = Field(
        default=None,
        max_length=50,
        description="摄像头编码"
    )
    description: Optional[str] = Field(
        default=None,
        description="摄像头描述"
    )
    area_id: Optional[str] = Field(
        default=None,
        description="所属区域ID"
    )
    rtsp_url: Optional[str] = Field(
        default=None,
        max_length=500,
        description="RTSP 流地址"
    )
    manufacturer: Optional[str] = Field(
        default=None,
        description="设备厂商"
    )
    device_model: Optional[str] = Field(
        default=None,
        description="设备型号"
    )
    ip_address: Optional[str] = Field(
        default=None,
        description="设备IP"
    )
    location: Optional[str] = Field(
        default=None,
        description="安装位置"
    )
    longitude: Optional[float] = Field(
        default=None,
        description="经度"
    )
    latitude: Optional[float] = Field(
        default=None,
        description="纬度"
    )
    fps: Optional[int] = Field(
        default=None,
        ge=1,
        le=60,
        description="帧率"
    )
    inference_interval_sec: Optional[int] = Field(
        default=None,
        ge=1,
        le=3600,
        description="识别间隔(秒)，控制该摄像头推理抽帧频率"
    )
    resolution: Optional[str] = Field(
        default=None,
        description="分辨率"
    )
    is_enabled: Optional[bool] = Field(
        default=None,
        description="是否启用"
    )


class CameraResponse(BaseModel):
    """
    摄像头响应
    """
    
    id: str = Field(description="摄像头ID")
    name: str = Field(description="摄像头名称")
    code: Optional[str] = Field(description="摄像头编码")
    description: Optional[str] = Field(description="摄像头描述")
    area_id: Optional[str] = Field(description="所属区域ID")
    area_name: Optional[str] = Field(default=None, description="所属区域名称")
    rtsp_url: str = Field(description="RTSP 流地址")
    manufacturer: Optional[str] = Field(description="设备厂商")
    device_model: Optional[str] = Field(description="设备型号")
    ip_address: Optional[str] = Field(description="设备IP")
    location: Optional[str] = Field(description="安装位置")
    longitude: Optional[float] = Field(description="经度")
    latitude: Optional[float] = Field(description="纬度")
    fps: int = Field(description="帧率")
    inference_interval_sec: int = Field(description="识别间隔(秒)")
    resolution: Optional[str] = Field(description="分辨率")
    is_enabled: bool = Field(description="是否启用")
    status: str = Field(description="在线状态")
    algorithm_count: int = Field(default=0, description="关联算法数量")
    created_at: datetime = Field(description="创建时间")
    updated_at: datetime = Field(description="更新时间")
    
    model_config = {"from_attributes": True}


class CameraListResponse(BaseModel):
    """
    摄像头列表响应
    """
    
    code: int = Field(default=0, description="状态码")
    message: str = Field(default="success", description="消息")
    data: List[CameraResponse] = Field(
        default_factory=list,
        description="摄像头列表"
    )
    page_info: Optional[dict] = Field(
        default=None,
        description="分页信息"
    )


class CameraStatusResponse(BaseModel):
    """
    摄像头状态响应
    """
    
    id: str = Field(description="摄像头ID")
    name: str = Field(description="摄像头名称")
    status: str = Field(description="状态: online/offline/error")
    is_enabled: bool = Field(description="是否启用")
    play_url: Optional[str] = Field(default=None, description="播放地址")
    
    model_config = {"from_attributes": True}


class CameraPlayUrlResponse(BaseModel):
    """
    摄像头播放地址响应
    """
    
    camera_id: str = Field(description="摄像头ID")
    flv_url: Optional[str] = Field(default=None, description="HTTP-FLV 播放地址")
    rtsp_url: Optional[str] = Field(default=None, description="RTSP 播放地址")
    hls_url: Optional[str] = Field(default=None, description="HLS 播放地址")
